### PR TITLE
[#262] Remove FA2 supports 

### LINKED ISSUE #262: Consider removing FA2 compliance from the contract
# Clarification and motivation

After the BYO changes, now the DAO relies on an external FA2 governance token, whereas in itself it only stores the froze token and possibly custom additional ones.

However the custom additional tokens are not used by the two main implementations (both `registryDAO` and `treasuryDAO`) and are likely not very useful to most custom DAOs (if necessary, one can always use other external/existing FA2 contracts, as for the governance one).

Moreover, taking in consideration that frozen tokens:
- cannot be `transfer`ed
- have to be stored in a `ledger` for `balance_of` and `update_operator` to work (for the FA2 compliance)
- have much custom logic that don't apply to other tokens (e.g. they have to remember in which stage the tokes were frozen to know in which they can be used)

this causes a lot of inefficiency, as the storage requires "normal" token updates for FA2 as well as "special" info for the custom logic.

Given that both `transfer` and `update_operator` (as a consequence) are not really usable, this seems a big and unnecessary inefficiency, which could perhaps be removed.

# Acceptance criteria

1. consider and discuss dropping FA2 support in the DAO itself
2. if an agreement is reached, make the necessary changes.

### PR DESCRIPTION
**Pull request by rinn7e**
## Description

Problem: After the BYO changes, now the DAO relies on an external
FA2 governance token, whereas in itself it only stores the froze
token and possibly custom additional ones.
However the custom additional tokens are not used by the two main
implementations (both `registryDAO` and `treasuryDAO`) and are likely
not very useful to most custom DAOs (if necessary, one can always use
other external/existing FA2 contracts, as for the governance one).

Solution: Remove FA2 entrypoints, remove `ledger` and `total_supply`,
handle all the affected changes, update the tests.

## Related issue(s)

Resolves #262 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
Problem: When a proposal is rejected, we burn its value in the ledger
and total_supply, but not in freeze_history.

Solution: Update `unstake_tk` to take `burn_amount` into account.

### CODE REVIEW

File: src/proposal.mligo
Comment (rinn7e) [inline src/proposal.mligo:221]: @pasqu4le Took me a while to update the tests, since there are a few strange cases coming up. But they should be all fine now.

One note is that, since we remove `total_supply`, I have to used `governance_total_supply` here. Not sure if this is to be expected.

This PR is pretty much ready for review, only the specification is left. I'll work on that next.

Comment (sras) [reply src/proposal.mligo:244]: @rinn7e I am a bit doubtful of this change that uses `governance_total_supply` for `frozen_total_supply`.

AFAIU, the governance total supply is the meant to be the total supply of the unfrozen tokens, some of which can be converted to frozen tokens. And the amount of frozen tokens for an account is considered as a measure of how much a voter wants to invest in the DAO.

Also, it does not make sense to use the governace_total_supply as a denominator for calculating the quorum, because AFAIU the denominator should be the maximum possible votes, which is related to the number of frozen tokens, aka, `frozen_total_supply`, and not related to the governance total supply.

@pasqu4le should probably  confirm.

Comment (rinn7e) [reply src/proposal.mligo:244]: I see, if this is the case, then we should not remove `total_supply`. Let me know if we should put this back in.

File: src/types.mligo
Comment (sras) [inline src/types.mligo:178]: @rinn7e I didn't get why we need to add the `from_` field here and for `vote_param`, can you please clarify?

File: haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs
Comment (sras) [inline haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs:83]: I get that we cannot check balance in network, but how does this help?

File: src/types.mligo
Comment (rinn7e) [reply src/types.mligo:178]: This is related to the new role `delegate` which is similar to FA2 `operator` but instead can call `propose` and `vote` entrypoints. Since we can't use `sender` anymore, we need a new field. More discussion here: https://serokell.slack.com/archives/C019PCGKTK2/p1622222486020100?thread_ts=1621943302.012700&cid=C019PCGKTK2

File: src/proposal/freeze_history.mligo
Comment (sras) [inline src/proposal/freeze_history.mligo:24]: May be we should add a test for this.

Comment (rinn7e) [reply src/proposal/freeze_history.mligo:24]: With the new `checkBalance` function, it already uses `freeze_history` to check the balance, which means it covers `unstake_frozen_fh` as well. https://github.com/tqtezos/baseDAO/pull/275/files#diff-d519aedaf6d7160f71739847f8cdfe0224fdee4cd758cd76e5d56df3e62215f3R86

Or maybe you have some other tests in mind, could suggest to me as well.

File: haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs:83]: This allows using the same test case, and run it in both network and emulator:

https://github.com/tqtezos/baseDAO/pull/275/files#diff-bce87f85bb8daa411eca6841102ce815e6bccc0b9977a22c827f936f12559810R124-R129

I could also just use `(\ _ _ _ -> pure ())`, but it is mainly for consistency purpose.

Comment (sras) [reply haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs:83]: @rinn7e But AFAIU it does not check anything, so wouldn't it make the test useless?

File: src/proposal/freeze_history.mligo
Comment (sras) [reply src/proposal/freeze_history.mligo:24]: I see. AFAIU, it now check the freeze balance by summing all the field in `address_freeze_history`, so cannot detect if two field values are swapped (because the sum will remain the same). May be it is good enough. I ll leave it up to you.

File: haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs:83]: Using this test as example: https://github.com/tqtezos/baseDAO/pull/275/files#diff-8ce82adc395e63130a569731f81e84d545f091f9038d4bb597c84aed02e0c221R28-R38

Basically, this test does a bunch of things including checkBalance. When running in network, it just skip the checkBalance, but it still run the rest of the actions.

Comment (sras) [reply haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs:83]: Ok, may be we should rename the function to something like "checkBalanceNetworkDummy", or use 

>  (\ _ _ _ -> pure ())

inline.

File: src/proposal.mligo
Comment (pasqu4le) [reply src/proposal.mligo:244]: Yeah, @sras is correct, we cannot use the `governance_total_supply` for that we need to store the `total_supply` indeed.

However keep in mind that we only need to store that for frozen tokens, so it can be a single `nat` in the storage, I suggest naming it `frozen_total_supply`.

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:60]: A comment above this function could be useful to tell what it does, for example without looking at the code it's not clear what address is returned.

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:50]: I think that now that `propose_params` includes the source address we don't need to provide it separately, AFAIU we could change `to_proposal_key` to only have one argument.

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:93]: 2 suggestions:

1. rename this to `lock_governance_tokens`, since there is no longer a ledger to freeze on and all that this function does now is indeed lock governance tokens into the DAO's account.
2. nitpick: return `operation list` instead (with the only element), it will make this cleaner to use in its only call site

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:117]: Same two suggestions here, except that I would name this `release_governance_tokens`, or maybe `unlock_governance_tokens`.

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:124]: Why is `from_` needed? Can't we use `propose_param.from_`?

File: src/proposal.mligo
Comment (pasqu4le) [inline src/proposal.mligo:205]: `store` is no longer needed here, since it doesn't get updated anymore.

File: haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Common/StorageHelper.hs]: I've updated this now.

File: Makefile
Comment (rinn7e) [reply Makefile]: Make sense, I've updated this makefile.

File: src/types.mligo
Comment (rinn7e) [reply src/types.mligo]: Updated.

File: src/proposal/freeze_history.mligo
Comment (rinn7e) [reply src/proposal/freeze_history.mligo]: Updated.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Make sense, I've updated this.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Added a comment on this.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: I've updated this now.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Renamed this as well as the `unlock_governance_tokens`

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Not really needed, I've removed this now.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Updated.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: I have added `frozen_total_supply` back now.

File: Makefile
Comment (pasqu4le) [inline Makefile:58]: @rinn7e the metadata generation need to be updated in the Haskell code.

Running `make metadata` now I see several things to remove:
1. `TZIP-12` is still listed as a supported interface
2. The `all_tokens` view, which is FA2-specific
3. The `token_metadata` view, which is also FA2-specific

File: docs/specification.md
Comment (pasqu4le) [inline docs/specification.md:25]: There is no `token_id` anymore.

File: docs/specification.md
Comment (pasqu4le) [inline docs/specification.md:198]: AFAIK the proposal states have not changed, what is `pending`?

File: docs/specification.md
Comment (pasqu4le) [inline docs/specification.md:214]: This shouldn't be in this list of "global" role, it's explanation I think should be more similar to the removed one for the `operator`

File: haskell/test/Test/Ligo/BaseDAO/Proposal/Delegate.hs
Comment (pasqu4le) [inline haskell/test/Test/Ligo/BaseDAO/Proposal/Delegate.hs:69]: A couple more things that would be nice to check:
1. that one can update multiple delegates
2. that (since it is a list) one can add and remove a delegate in the same call without an error
3. that one can remove a delegate even if it was not set before (since we are permissive about that)

File: docs/specification.md
Comment (sras) [inline docs/specification.md:18]: >ofs

Possible typo

File: docs/specification.md
Comment (sras) [inline docs/specification.md:215]: >partuclar

Possible typo

File: docs/specification.md
Comment (rinn7e) [reply docs/specification.md:198]: Seems like the previous proposal state are out dated. I'm using `pending` here to refer to "not yet voting period", probably have to reword this, hope to get your suggestion as well.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Updated the code as well as the doc.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Fixed.

File: src/proposal.mligo
Comment (rinn7e) [reply src/proposal.mligo]: Updated.

File: Makefile
Comment (rinn7e) [reply Makefile:58]: I forgot about this, pushed an update now.

File: docs/specification.md
Comment (rinn7e) [reply docs/specification.md]: Removed the `token_id` part.

File: docs/specification.md
Comment (rinn7e) [reply docs/specification.md]: Make sense, I've updated this.

File: haskell/test/Test/Ligo/BaseDAO/Proposal/Delegate.hs
Comment (rinn7e) [reply haskell/test/Test/Ligo/BaseDAO/Proposal/Delegate.hs:72]: Added a new test that covers all this cases.

File: docs/specification.md
Comment (rinn7e) [reply docs/specification.md]: Fixed.

File: docs/specification.md
Comment (pasqu4le) [reply docs/specification.md:198]: I don't think they were really outdated, from their creation onward proposals go through these states:
1. "ongoing" when they have been proposed but can't be voted on yet (I guess we can rename it to "proposed")
2. "ongoing" when it's getting voted on
3. "ongoing" if it's no longer time to vote on it, but it hasn't been flushed yet (I guess this can be called "pending flush")
3. "accepted" or "rejected" when they are flushed, depending on the result
4. "expired" if they haven't been flushed in time

If you agree I think we can make this more clear here.

**Review by pasqu4le — APPROVED** [review]
needs a rebase , but otherwise LGTM now

File: docs/specification.md
Comment (rinn7e) [reply docs/specification.md]: I've updated it as the follow:
```
- The contract stores a list of proposals that can be in one of the states:
  "proposed", "ongoing", "pending flush", "accepted, "rejected", or "expired".
```

Let me know if this is good enough. I also rebased and squashed the commits now.

Comment (pasqu4le) [reply docs/specification.md]: Yep, sounds good to me, except that you forgot a double quote after `accepted`. it should be:
```
"proposed", "ongoing", "pending flush", "accepted", "rejected", or "expired".
```

The rebase also looks good to me :+1:

**Pull request merged**
